juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1555 @@
+use std::collections::HashMap;
+
+use crate::{
+    ast::{Arguments, Directive, Field, Fragment, InputValue, Selection},
+    parser::{Span, Spanning},
+    value::ScalarValue,
+};
+
+use super::Variables;
+
+/// An enum that describes if a field is available in all types of the interface
+/// or only in a certain subtype
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Applies<'a> {
+    /// The field is available independent from the type
+    All,
+    /// The field is only available for a given typename
+    OnlyType(&'a str),
+}
+
+/// Shortcut for a [`Spanning`] containing a borrowed [`Span`].
+type BorrowedSpanning<'a, T> = Spanning<T, &'a Span>;
+
+/// JSON-like value that can be used as an argument in the query execution.
+///
+/// In contrast to an [`InputValue`], these values do only contain constants,
+/// meaning that variables get automatically resolved.
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(missing_docs)]
+pub enum LookAheadValue<'a, S: ScalarValue + 'a> {
+    Null,
+    Scalar(&'a S),
+    Enum(&'a str),
+    List(LookAheadList<'a, S>),
+    Object(LookAheadObject<'a, S>),
+}
+
+impl<'a, S: ScalarValue + 'a> LookAheadValue<'a, S> {
+    fn from_input_value(
+        input_value: &'a InputValue<S>,
+        span: &'a Span,
+        vars: Option<&'a Variables<S>>,
+    ) -> BorrowedSpanning<'a, Self> {
+        Spanning {
+            item: match input_value {
+                InputValue::Null => Self::Null,
+                InputValue::Scalar(s) => Self::Scalar(s),
+                InputValue::Enum(e) => Self::Enum(e),
+                InputValue::Variable(name) => vars
+                    .and_then(|vars| vars.get(name))
+                    .map(|item| Self::from_input_value(item, span, vars).item)
+                    .unwrap_or(Self::Null),
+                InputValue::List(input_list) => Self::List(LookAheadList { input_list, vars }),
+                InputValue::Object(input_object) => Self::Object(LookAheadObject {
+                    input_object: input_object.as_slice(),
+                    vars,
+                }),
+            },
+            span,
+        }
+    }
+}
+
+/// A JSON-like list that can be used as an argument in the query execution.
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadList<'a, S> {
+    input_list: &'a [Spanning<InputValue<S>>],
+    vars: Option<&'a Variables<S>>,
+}
+
+impl<'a, S: ScalarValue> LookAheadList<'a, S> {
+    /// Returns an iterator over the list's elements.
+    pub fn iter(&self) -> impl Iterator<Item = BorrowedSpanning<'a, LookAheadValue<'a, S>>> + '_ {
```

Comment:
These wrapper types should ideally have `IntoIterator` implementations I suppose, but I don't think the added boilerplate justifies it currently. I think it can maybe be added later when/if Rust gets the ability to write shorter IntoIterator implementations. Especially I think exposing a new public type for every iterator is a little painful.

Reply:
Done.